Sources/SemanticIndex/SemanticIndexManager.swift

```diff
@@ -0,0 +1,167 @@
+    self.buildSystemManager = buildSystemManager
+    self.indexTaskScheduler = indexTaskScheduler
+    self.indexTaskDidFinish = indexTaskDidFinish
+  }
+
+  /// Schedules a task to index all files in `files` that don't already have an up-to-date index.
+  /// Returns immediately after scheduling that task.
+  ///
+  /// Indexing is being performed with a low priority.
+  public func scheduleBackgroundIndex(files: some Collection<DocumentURI>) {
+    self.index(files: files, priority: .low)
+  }
+
+  /// Wait for all in-progress index tasks to finish.
+  public func waitForUpToDateIndex() async {
+    logger.info("Waiting for up-to-date index")
+    await withTaskGroup(of: Void.self) { taskGroup in
+      for (_, status) in indexStatus {
+        switch status {
+        case .inProgress(let task):
+          taskGroup.addTask {
+            await task.value
+          }
+        case .upToDate:
+          break
+        }
+      }
+      await taskGroup.waitForAll()
+    }
+    index.pollForUnitChangesAndWait()
+    logger.debug("Done waiting for up-to-date index")
+  }
+
+  /// Ensure that the index for the given files is up-to-date.
+  ///
+  /// This tries to produce an up-to-date index for the given files as quickly as possible. To achieve this, it might
+  /// suspend previous target-wide index tasks in favor of index tasks that index a fewer files.
+  public func waitForUpToDateIndex(for uris: some Collection<DocumentURI>) async {
+    logger.info(
+      "Waiting for up-to-date index for \(uris.map { $0.fileURL?.lastPathComponent ?? $0.stringValue }.joined(separator: ", "))"
+    )
+    let filesWithOutOfDateIndex = uris.filter { uri in
+      switch indexStatus[uri] {
+      case .inProgress, nil: return true
+      case .upToDate: return false
+      }
+    }
+    // Create a new index task for the files that aren't up-to-date. The newly scheduled index tasks will
+    // - Wait for the existing index operations to finish if they have the same number of files.
+    // - Reschedule the background index task in favor of an index task with fewer source files.
+    await self.index(files: filesWithOutOfDateIndex, priority: nil).value
+    index.pollForUnitChangesAndWait()
+    logger.debug("Done waiting for up-to-date index")
+  }
+
+  // MARK: - Helper functions
+
+  /// Index the given set of files at the given priority.
+  ///
+  /// The returned task finishes when all files are indexed.
+  @discardableResult
+  private func index(files: some Collection<DocumentURI>, priority: TaskPriority?) -> Task<Void, Never> {
+    let outOfDateFiles = files.filter {
+      if case .upToDate = indexStatus[$0] {
+        return false
+      }
+      return true
+    }
+
+    var indexTasks: [Task<Void, Never>] = []
+
+    // TODO (indexing): Group index operations by target when we support background preparation.
+    for files in outOfDateFiles.partition(intoNumberOfBatches: ProcessInfo.processInfo.processorCount * 5) {
```

Comment:
Worth a comment explaining the batching size? Or does it not matter since this will change once we support preparation?

Reply:
I am actually already changing this to one index file per task in https://github.com/apple/sourcekit-lsp/pull/1273. 

https://github.com/apple/sourcekit-lsp/issues/1268 is when the batching really starts to make sense.